payments-core/src/main/java/com/stripe/android/payments/core/authentication/Stripe3DS2Authenticator.kt

```diff
@@ -300,59 +310,46 @@ internal class Stripe3DS2Authenticator(
         transaction: Transaction,
         sourceId: String,
         maxTimeout: Int,
-        paymentRelayStarter: PaymentRelayStarter,
-        host: AuthActivityStarter.Host,
+        host: AuthActivityStarterHost,
         stripeIntent: StripeIntent,
         requestOptions: ApiRequest.Options
     ) = withContext(workContext) {
-        runCatching {
-            requireNotNull(
-                host.fragment?.let { fragment ->
-                    Stripe3ds2ActivityStarterHost(fragment)
-                } ?: host.activity?.let { activity ->
-                    Stripe3ds2ActivityStarterHost(activity)
-                }
-            ) {
-                "Error while attempting to start 3DS2 challenge flow."
+        when (host) {
+            is AuthActivityStarterHost.ActivityHost -> {
+                Stripe3ds2ActivityStarterHost(host.activity)
             }
-        }.fold(
-            onSuccess = { stripe3ds2Host ->
-                delay(StripePaymentController.CHALLENGE_DELAY)
+            is AuthActivityStarterHost.FragmentHost -> {
+                Stripe3ds2ActivityStarterHost(host.fragment)
+            }
+        }.let { stripe3ds2Host ->
+            delay(StripePaymentController.CHALLENGE_DELAY)
 
-                transaction.doChallenge(
-                    stripe3ds2Host,
-                    ChallengeParameters(
-                        acsSignedContent = ares.acsSignedContent,
-                        threeDsServerTransactionId = ares.threeDSServerTransId,
-                        acsTransactionId = ares.acsTransId
-                    ),
-                    DefaultStripeChallengeStatusReceiver(
-                        stripe3ds2CompletionStarterFactory(
-                            host,
-                            StripePaymentController.getRequestCode(stripeIntent)
-                        ),
-                        stripeRepository,
-                        stripeIntent,
-                        sourceId,
-                        requestOptions,
-                        analyticsRequestExecutor,
-                        analyticsRequestFactory,
-                        transaction,
-                        {
-                            transaction.close()
-                        },
-                        workContext = workContext
+            transaction.doChallenge(
+                stripe3ds2Host,
+                ChallengeParameters(
+                    acsSignedContent = ares.acsSignedContent,
+                    threeDsServerTransactionId = ares.threeDSServerTransId,
+                    acsTransactionId = ares.acsTransId
+                ),
+                DefaultStripeChallengeStatusReceiver(
+                    stripe3ds2CompletionStarterFactory(
+                        host,
+                        StripePaymentController.getRequestCode(stripeIntent)
                     ),
-                    maxTimeout
-                )
-            },
-            onFailure = {
```

Comment:
hm, looking for where the failure handling went.

Reply:
It's a bit hard to follow, but you can see that the `runCatching` block at the top of this method was removed. This is because the components are lifecycle-aware.